text/0000-block-manager.md

```diff
@@ -0,0 +1,367 @@
+    corresponding `unref_block()` with the same block id is called.
+
+`unref_block(block_id: String)`
+
+    Release a previous `ref_block()` on a block. May result in the block being
+    removed from the tree.
+
+    Raises an exception if the reference count falls below 0, which indicates
+    a bug in the application code.
+
+The following example illustrates how the `ref_block` operation is used to
+prevent blocks from being dropped when they are transferred from the completer
+to the chain controller.
+
+Assume that the chain controller has a block A and that the completer has just
+completed a block B and would like to pass it to the chain controller. If the
+completer were to just pass the chain controller B through a queue or shared
+memory, the chain controller could decide to `unref_block` A before it receives
+B, causing B to have a missing predecessor when it arrives at the chain
+controller. Instead, the completer first places a `ref_block` on A, causing and
+then it passes B to the chain controller. If the chain controller decides to
+`unref_block` A now, the block manage will ensure it is not actually dropped,
+since there is still an open hold on the block. Finally, when the chain
+controller takes ownership of B, it can send a signal back to the completer
+that it has been received and the completer can release its hold.
+
+### Persisting Blocks
+
+In order to support persisting blocks to an alternative not-in-memory location,
+a generic __block store__ interface is defined and the __block manager__
+supports adding any number of block stores and transferring ownership of the
+blocks to these stores.
+
+Persisting blocks is handled by the following methods:
+
+`add_store(store_name: String, store: Store)`
+
+    Take ownership of the given store and enable it for persisting blocks. The
+    store must be referenced from other methods with `store_name`.
+
+    Raises an exception if a store with `store_name` already exists.
+
+`persist(head: String, store_name: String)`
+
+    Atomically ensure that all blocks on the branch starting with head are
+    in the store.
+
+## Block Store Interface
+
+The __block store__ interface is the interface that must be implemented by an
+object in order for it to be added to the block tree as a "store". The
+following is a listing of the interface's methods:
+
+`put(blocks: List<Block>)`
+
+    Atomically add the given blocks to the store.
+
+`get(block_ids: List<String>) -> Iter<Block>`
+
+    Atomically get the blocks with the given ids.
+
+`iter() -> Iter<Block>`
+
+    Create an iterator over all blocks in the store.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following provides more detail on how the API present above is implemented.
+
+## Internal Representation
+
+Internally, the block manager consists of:
+
+- A __main cache__ which, at a minimum, contains all blocks that have not been
```

Comment:
is main cache a block store?

Reply:
Hi, @dcmiddle . Since I am implementing this right now, I thought I should respond. The main cache is in-memory and is not a block store. It is a logical space. Right now in my implementation it is a HashMap of block ids to Block - reference count tuples, wrapped in a lock. Specifically Rc<RwLock<HashMap<String, (Rc<Block>, i64, i64)>>>.